fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OuterJoinAddRedundantTopNRule.java

```diff
@@ -0,0 +1,100 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.collect.Lists;
+import com.starrocks.analysis.JoinOperator;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.Ordering;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalTopNOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class OuterJoinAddRedundantTopNRule extends TransformationRule {
+    public OuterJoinAddRedundantTopNRule() {
+        super(RuleType.TF_PUSH_DOWN_TOPN_OUTER_JOIN,
+                Pattern.create(OperatorType.LOGICAL_TOPN).addChildren(Pattern.create(OperatorType.LOGICAL_JOIN)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalTopNOperator topn = (LogicalTopNOperator) input.getOp();
+
+        OptExpression childExpr = input.inputAt(0);
+        LogicalJoinOperator joinOperator = childExpr.getOp().cast();
+        JoinOperator joinType = joinOperator.getJoinType();
+
+        if (!joinType.isLeftOuterJoin() && !joinType.isRightOuterJoin()) {
+            return false;
+        }
+
+        List<Integer> colIds = topn.getOrderByElements().stream()
+                .map(Ordering::getColumnRef)
+                .map(ColumnRefOperator::getId)
+                .collect(Collectors.toList());
+        if (joinType.isLeftOuterJoin()) {
+            OptExpression joinLeftChild = childExpr.inputAt(0);
+            return joinLeftChild.getOutputColumns().containsAll(colIds);
+        }
+
+        if (joinType.isRightJoin()) {
```

Comment:
you should consider the colocate aggregate + colocate join

Reply:
I have controlled the default value of limit to 1000, and I think that in most cases, the benefit of pushdown top is greater than that of colocate join. And I only push one layer of topn nodes, and the gather nodes will not affect too much. If there is a bad case, user can control the limit by the session variable to skip the rule.